addon development guide (#628)
harvester/docs: docs/developer/addon-development.md

---
id: addon-development-guide
title: "Addon Development Guide"
sidebar_position: 1
sidebar_label: Addon Development Guide
keywords:
  - Harvester
  - harvester
  - Addon Development

> Description: How to write your own Harvester Addon

Reviewer: "addon" is not a word. It is "Add-on" and I'm working on a PR to actively reflect that: https://github.com/harvester/dashboard/pull/1110 Please adjust accordingly

---

# Background
Harvester Addons provide a means to enable/disable specific Harvester or third party components based on the user requirements. 

Addons provide a wrapper around [RKE2 HelmChart CRD](https://docs.rke2.io/helm#using-the-helm-crd).

## Prerequisites

- An existing Harvester cluster

## Addon Spec

The Addon CRD supports the following fields

```yaml
apiVersion: harvesterhci.io/v1beta1
kind: Addon                         
metadata:
  name: example-addon           # name of Addon
  namespace: example-namespace  # namespace where Addon is deployed. The associated k8s components will be deployed in the same namespace as the Addon crd

>   labels:                       # optional aAddon labels

Reviewer 2: ```suggestion
  labels:                       # optional Addon labels
```

    addon.harvesterhci.io/experimental: "true" # predefined label used by Harvester dashboard to indicate experimental tag in UI
spec:

>   enabled: false                # boolean indicating if aAddon should be enabled or disabled on definition

Reviewer 2: ```suggestion
  enabled: false                # boolean indicating if an Addon should be enabled or disabled on definition
```

Not 100% sure about this one, but looks like a typo to me.

  repo: https://chartsrepo.com  # helm chart repo containing the helm chart being managed by Addon
  version: "v0.19.0"            # version of helm chart to be installed
  chart: mychart                # name of helm chart in the helm chart repo

>   valuesContent: |-             # values.yaml that needs to be passed to the helm chart

Reviewer 2: Should we also let users know it's possible to template values like the vcluster add-on? https://github.com/harvester/experimental-addons/blob/main/rancher-vcluster/rancher-vcluster.yaml

Author: This is only possible because of templating being supported in vcluster itself. It will not work with most charts.

---
id: Add-on-development-guide
title: "Add-on Development Guide"
sidebar_position: 1
sidebar_label: Add-on Development Guide
keywords:
  - Harvester
  - harvester
  - Add-on Development
Description: How to write your own Harvester Add-on
---

> # Background
> Harvester Add-ons provide a means to enable/disable specific Harvester or third party components based on the user requirements. 
> 
> Add-ons provide a wrapper around [RKE2 HelmChart CRD](https://docs.rke2.io/helm#using-the-helm-crd).

Reviewer 3: ```suggestion
Harvester add-ons allow you to enable and disable specific Harvester and third-party components based on your requirements. Add-ons function as a wrapper for the [RKE2 HelmChart resource definition (CRD)](https://docs.rke2.io/helm#using-the-helm-crd).
```

Reviewer 3: The heading is not necessary. Also, H1 is reserved for the main title.

## Prerequisites

- An existing Harvester cluster

> ## add-on Spec
> 
> The `Addon` CRD supports the following fields

Reviewer 3: ```suggestion
## Add-on Specifications

The `Addon` CRD supports the following fields:
```

```yaml
apiVersion: harvesterhci.io/v1beta1
kind: Addon                         
metadata:

>   name: example-add-on           # name of add-on
>   namespace: example-namespace  # namespace where add-on is deployed. The associated k8s components will be deployed in the same namespace as the addon crd
>   labels:                       # optional add-on labels
>     Add-on.harvesterhci.io/experimental: "true" # predefined label used by Harvester dashboard to indicate experimental tag in UI

Reviewer 3: ```suggestion
  name: example-add-on           # Name of add-on
  namespace: example-namespace  # Namespace where the add-on CRD is deployed and where the associated Kubernetes components will be deployed
  labels:                       # Optional add-on labels
    Add-on.harvesterhci.io/experimental: "true" # Predefined label used on the Harvester UI to identify add-ons with the "experimental" maturity level
```

spec:

>   enabled: false                # boolean indicating if an add-on should be enabled or disabled on definition
>   repo: https://chartsrepo.com  # helm chart repo containing the helm chart being managed by add-on
>   version: "v0.19.0"            # version of helm chart to be installed
>   chart: mychart                # name of helm chart in the helm chart repo
>   valuesContent: |-             # values.yaml that needs to be passed to the helm chart

Reviewer 3: ```suggestion
  enabled: false                # Boolean indicating if an add-on should be enabled or disabled on definition
  repo: https://chartsrepo.com  # Helm chart repository containing the Helm chart managed by the add-on
  version: "v0.19.0"            # Version of the Helm chart to be installed
  chart: mychart                # Name of the Helm chart in the Helm chart repository
  valuesContent: |-             # File (values.yaml) that must be passed to the Helm chart
```

    contents of values.yaml
    that need to be passed
    to the chart
```

> An example add-on is the [rancher-vluster add-on.](https://github.com/harvester/experimental-addons/blob/main/rancher-vcluster/rancher-vcluster.yaml)

Reviewer 3: ```suggestion
:::note

Experimental add-ons are not directly packaged in Harvester. An example is the [rancher-vcluster](https://github.com/harvester/experimental-addons/blob/main/rancher-vcluster/rancher-vcluster.yaml) add-on.

:::
```

Reviewer 3: I added a sentence to provide more context and connect the note to the preceding code block.

## Installation

```bash
kubectl apply -f /path/to/add-on.yaml
```

> ## Usage
> Once an `Addon` CRD has been created, end users can toggle the `enabled` field to `true/false` to accordingly enable and disable the associated Helm Chart
> 
> ## Upgrade
> Changes to the `repo`, `version`, `chart` or `valueContent` field will trigger an underlying `helm upgrade` which will force the existing helm chart to be upgraded

Reviewer 3: ```suggestion
## Usage

After creating an `Addon` CRD, you can can toggle the `enabled` field to enable and disable the associated Helm chart.

## Upgrade

Changes to the `repo`, `version`, `chart` or `valueContent` fields will trigger a `helm upgrade`, which forces an upgrade of the existing Helm chart.
```

---
id: Add-on-development-guide
title: "Add-on Development Guide"
sidebar_position: 1
sidebar_label: Add-on Development Guide
keywords:
  - Harvester
  - harvester
  - Add-on Development

> Description: How to write your own Harvester Add-on

Reviewer: ```suggestion
Description: How to write your own Harvester add-on
